File: backend/emw_backend/routes/society.py
from __future__ import annotations

import uuid
from typing import Optional

from flask import Blueprint, current_app, jsonify, request
from sqlalchemy import desc, select

from emw_backend.auth import require_auth_user
from emw_backend.db import SessionLocal
from emw_backend.models import SocietyComment, SocietyPost, UserDevice

society_bp = Blueprint("society", __name__, url_prefix="/v1/society")


def _now_ms() -> int:
    import time

    return int(time.time() * 1000)


def _parse_int(name: str, raw: Optional[str], default: int, min_v: int, max_v: int) -> int:
    try:
        v = int(raw) if raw is not None else int(default)
    except Exception:
        v = int(default)
    if v < min_v:
        v = min_v
    if v > max_v:
        v = max_v
    return v


@society_bp.get("/posts")
def list_posts():
    """Read-only list of EMWaver Society posts.

    Query:
      kind: announcement|discussion|script|video (optional)
      limit: int (default 20)
      before_ms: cursor for pagination (created_at_ms < before_ms)
    """

    kind = (request.args.get("kind") or "").strip().lower() or None
    limit = _parse_int("limit", request.args.get("limit"), default=20, min_v=1, max_v=50)

    before_ms_raw = (request.args.get("before_ms") or "").strip()
    before_ms = None
    if before_ms_raw:
        try:
            before_ms = int(before_ms_raw)
        except Exception:
            before_ms = None

    db = SessionLocal()
    try:
        q = select(SocietyPost).where(SocietyPost.published == 1)
        if kind:
            q = q.where(SocietyPost.kind == kind)
        if before_ms is not None:
            q = q.where(SocietyPost.created_at_ms < before_ms)

        q = q.order_by(desc(SocietyPost.created_at_ms)).limit(limit)
        rows = db.execute(q).scalars().all()

        return jsonify(
            {
                "posts": [r.to_public_dict() for r in rows],
            }
        )
    finally:
        db.close()


@society_bp.get("/posts/<post_id>")
def get_post(post_id: str):
    db = SessionLocal()
    try:
        row = db.get(SocietyPost, post_id)
        if not row or not row.published:
            return jsonify({"error": "not_found"}), 404
        return jsonify({"post": row.to_public_dict(include_body=True)})
    finally:
        db.close()


@society_bp.get("/posts/<post_id>/comments")
def list_comments(post_id: str):
    limit = _parse_int("limit", request.args.get("limit"), default=50, min_v=1, max_v=200)

    db = SessionLocal()
    try:
        post = db.get(SocietyPost, post_id)
        if not post or not post.published:
            return jsonify({"error": "not_found"}), 404

        q = (
            select(SocietyComment)
            .where(SocietyComment.post_id == post_id)
            .order_by(SocietyComment.created_at_ms.asc())
            .limit(limit)
        )
        rows = db.execute(q).scalars().all()
        return jsonify({"comments": [c.to_public_dict() for c in rows]})
    finally:
        db.close()


@society_bp.post("/posts/<post_id>/comments")
def create_comment(post_id: str):
    """Create a comment.

    Requirements:
      - signed in
      - account has >=1 attached/verified genuine device

    Payload: { body_md }
    """

    config = current_app.config.get("EMWAVER_CONFIG")
    user = require_auth_user(config)

    payload = request.get_json(silent=True) or {}
    body_md = str(payload.get("body_md") or "").strip()
    if not body_md:
        return jsonify({"error": "missing_body_md"}), 400
    if len(body_md) > 20_000:
        return jsonify({"error": "body_too_long"}), 400

    db = SessionLocal()
    try:
        post = db.get(SocietyPost, post_id)
        if not post or not post.published:
            return jsonify({"error": "not_found"}), 404
        if post.locked:
            return jsonify({"error": "locked"}), 403

        # Device-attached gating (anti-spam + aligns with platform policy).
        has_device = db.execute(
            select(UserDevice.device_id_b64).where(UserDevice.firebase_uid == user.firebase_uid).limit(1)
        ).first()
        if not has_device:
            return jsonify({"error": "device_required"}), 403

        now = _now_ms()
        row = SocietyComment(
            id=str(uuid.uuid4()),
            post_id=post_id,
            firebase_uid=user.firebase_uid,
            author_email=user.email,
            author_display_name=user.display_name,
            body_md=body_md,
            created_at_ms=now,
            updated_at_ms=now,
        )
        db.add(row)

        # Bump post updated time for activity ordering later.
        post.updated_at_ms = now

        db.commit()
        return jsonify({"comment": row.to_public_dict()}), 201
    finally:
        db.close()

File: web/src/server/store/society.ts
import { randomUUID } from "node:crypto";

import { readCollection, writeCollection } from "./jsonStore";

export type SocietyPostRecord = {
  id: string;
  kind: string;
  title: string;
  summary: string;
  body_md: string;
  firebase_uid: string | null;
  author_email: string | null;
  author_display_name: string | null;
  published: boolean;
  pinned: boolean;
  locked: boolean;
  pro_only?: boolean;
  created_at_ms: number;
  updated_at_ms: number;
};

export type SocietyCommentRecord = {
  id: string;
  post_id: string;
  firebase_uid: string;
  author_email: string | null;
  author_display_name: string | null;
  body_md: string;
  created_at_ms: number;
  updated_at_ms: number;
};

type SocietyState = {
  posts: Record<string, SocietyPostRecord>;
  comments: Record<string, SocietyCommentRecord>;
};

function nowMs() {
  return Date.now();
}

function defaultState(): SocietyState {
  return { posts: {}, comments: {} };
}

class SocietyStore {
  private posts = new Map<string, SocietyPostRecord>();
  private comments = new Map<string, SocietyCommentRecord>();

  constructor() {
    const state = readCollection<SocietyState>("society", defaultState());
    this.posts = new Map(Object.entries(state.posts || {}));
    this.comments = new Map(Object.entries(state.comments || {}));
  }

  private persist() {
    writeCollection("society", {
      posts: Object.fromEntries(this.posts.entries()),
      comments: Object.fromEntries(this.comments.entries()),
    });
  }

  listPosts(input: { kind?: string | null; before_ms?: number | null; limit: number }) {
    return [...this.posts.values()]
      .filter((post) => post.published)
      .filter((post) => !input.kind || post.kind === input.kind)
      .filter((post) => input.before_ms == null || post.created_at_ms < input.before_ms)
      .sort((a, b) => {
        if (a.pinned !== b.pinned) return Number(b.pinned) - Number(a.pinned);
        return b.created_at_ms - a.created_at_ms;
      })
      .slice(0, input.limit);
  }

  getPost(id: string) {
    return this.posts.get(id) || null;
  }

  listComments(postId: string, limit: number) {
    return [...this.comments.values()]
      .filter((comment) => comment.post_id === postId)
      .sort((a, b) => a.created_at_ms - b.created_at_ms)
      .slice(0, limit);
  }

  createComment(input: {
    post_id: string;
    firebase_uid: string;
    author_email: string | null;
    author_display_name: string | null;
    body_md: string;
  }) {
    const now = nowMs();
    const comment: SocietyCommentRecord = {
      id: randomUUID(),
      post_id: input.post_id,
      firebase_uid: input.firebase_uid,
      author_email: input.author_email,
      author_display_name: input.author_display_name,
      body_md: input.body_md,
      created_at_ms: now,
      updated_at_ms: now,
    };
    this.comments.set(comment.id, comment);
    const post = this.posts.get(input.post_id);
    if (post) {
      post.updated_at_ms = now;
      this.posts.set(post.id, post);
    }
    this.persist();
    return comment;
  }

  createThread(input: {
    title: string;
    summary: string;
    body_md: string;
    firebase_uid: string;
    author_email: string | null;
    author_display_name: string | null;
  }) {
    const now = nowMs();
    const post: SocietyPostRecord = {
      id: randomUUID(),
      kind: "discussion",
      title: input.title,
      summary: input.summary,
      body_md: input.body_md,
      firebase_uid: input.firebase_uid,
      author_email: input.author_email,
      author_display_name: input.author_display_name,
      published: true,
      pinned: false,
      locked: false,
      created_at_ms: now,
      updated_at_ms: now,
    };
    this.posts.set(post.id, post);
    this.persist();
    return post;
  }
}

const globalStore = globalThis as typeof globalThis & {
  __emwaverSocietyStore?: SocietyStore;
};

export const societyStore = globalStore.__emwaverSocietyStore ?? new SocietyStore();
globalStore.__emwaverSocietyStore = societyStore;
